src/data/fun_like/equiv.lean

```diff
@@ -115,36 +116,47 @@ This means anything set up for `my_iso`s will automatically work for `cool_iso_c
 and defining `cool_iso_class` only takes a constant amount of effort,
 instead of linearly increasing the work per `my_iso`-related declaration.
 
+If you already have a definition of homomorphisms (e.g. `my_hom_class`), you
+could also try defining the corresponding isomorphism by extending both
+`my_hom_class` and `equiv_like`. Make sure however that this does not create
+extra work in this particular instance: it's often easier to show a bijective
+map is an isomorphism, compared to showing a map is a homomorphism.  Compare the
+class of group homomorphisms `monoid_hom_class` which needs a proof of `map_mul`
+and `map_one`, and the class of group isomorphisms `mul_hom_class` which only
+needs `map_mul`. Thus, `mul_hom_class` does not extend `monoid_hom_class`.
 -/
 
+set_option old_structure_cmd true
+
 /-- The class `equiv_like E α β` expresses that terms of type `E` have an
 injective coercion to bijections between `α` and `β`.
 
-This typeclass is used in the definition of the homomorphism typeclasses,
+This typeclass is used in the definition of the isomorphism typeclasses,
 such as `zero_equiv_class`, `mul_equiv_class`, `monoid_equiv_class`, ....
 -/
-class equiv_like (E : Sort*) (α β : out_param Sort*) :=
-(coe : E → α → β)
+class equiv_like (E : Sort*) (α β : out_param Sort*) extends fun_like E α (λ _, β) :=
 (inv : E → β → α)
-(left_inv  : ∀ e, function.left_inverse (inv e) (coe e))
-(right_inv : ∀ e, function.right_inverse (inv e) (coe e))
--- The `inv` hypothesis makes this easier to prove with `congr'`
-(coe_injective' : ∀ e g, coe e = coe g → inv e = inv g → e = g)
+-- the `show` below fixes elaboration problems when using these fields stemming from the fact
+-- that in this case, `fun_like.coe : E → Π (a : α), (λ _, β) a`, not `E → α → β`.
+(left_inv  : ∀ e, function.left_inverse (inv e) (show α → β, from coe e))
+(right_inv : ∀ e, function.right_inverse (inv e) (show α → β, from coe e))
+(coe_inv_injective' : ∀ e g, coe e = coe g → inv e = inv g → e = g)
+(coe_injective' := λ e g h, coe_inv_injective' e g h $ funext $
+  λ x, left_inv g (inv e x) ▸ congr_arg (inv g) (h ▸ right_inv e x))
 
 namespace equiv_like
 
 variables {E F α β γ : Sort*} [iE : equiv_like E α β] [iF : equiv_like F β γ]
 include iE
 
 lemma inv_injective : function.injective (equiv_like.inv : E → (β → α)) :=
-λ e g h, coe_injective' e g ((right_inv e).eq_right_inverse (h.symm ▸ left_inv g)) h
+λ e g h, coe_injective' $ (right_inv e).eq_right_inverse (h.symm ▸ left_inv g)
 
 @[priority 100]
 instance to_embedding_like : embedding_like E α β :=
```

Comment:
After you define this instance, it's probably a good idea to make `to_fun_like` a non-instance; this type of "shortcut" inheritance can lead to performance issues.